BLOOM Flax

younesbelkada opened the pull request:
# What does this PR do?

An attempt of adding Flax implementation of BLOOM - original PR from @haileyschoelkopf #17761 

## TODOs:

- [x] alibi shifting for batched generation
- [ ] change mask fill value cc @patil-suraj 
- [x] optimize code (alibi creation + mask creation)

younesbelkada committed:
First commit

younesbelkada committed:
step 1 working

younesbelkada committed:
add alibi

Commit:
placeholder for `scan`

younesbelkada committed:
add matrix mult alibi

Commit:
beta scaling factor for bmm

younesbelkada committed:
working v1 - simple forward pass

Commit:
move layer_number from attribute to arg in call

Commit:
partial functioning scan

younesbelkada committed:
add more modifs

younesbelkada committed:
add test

Commit:
hacky working scan

Commit:
Merge remote-tracking branch 'younesbelkada/add_bloom_flax' into add_bloom_flax

Commit:
update scan for new kwarg order

patrickvonplaten committed:
Merge branch 'main' of https://github.com/huggingface/transformers into add_bloom_flax

patrickvonplaten committed:
fix position_ids problem

patrickvonplaten committed:
fix bug in attention layer

patrickvonplaten commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -379,25 +383,18 @@ def __call__(
-        # [q_length, batch_size, num_heads, head_dim] --> [q_length, batch_size, hidden_size]
-        new_context_layer_shape = context.shape[:-2] + (self.hidden_size,)
-        context = context.reshape(new_context_layer_shape)
+        attn_output = jnp.einsum("...hqk,...khd->...qhd", attention_probs, value)
```
@younesbelkada let's try to keep it simple - no need for all this manual reshaping. Our existing Flax models are quite optimized and we should follow the pattern that is already used in other models. So instead of copying from BLOOM PyTorch we should copy from OPT Flax

sanchit-gandhi replied:
Resolved in https://github.com/huggingface/transformers/pull/18022/commits/3444985451749540bd862910ebe7d01da8413771

patrickvonplaten commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -308,7 +309,10 @@ def __call__(
         output_attentions: bool = False,
         layer_number: int = None,
     ):
-        alibi = jnp.repeat(alibi, hidden_states.shape[0], axis=0)
```
this was a bug -> `jnp.repeat` interleaves instead of repeating for the batch

younesbelkada committed:
small fix

- do the alibi broadcasting only once

younesbelkada committed:
alibi shifting

Commit:
prelim refactor

Commit:
finish refactor

Commit:
Merge remote-tracking branch 'younesbelkada/add_bloom_flax' into add_bloom_flax

# Conflicts:
#	src/transformers/models/bloom/modeling_flax_bloom.py

Commit:
incorporate dropout_add to attention module

patrickvonplaten committed:
make style

patrickvonplaten committed:
make padding work again

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,826 @@
+
+def shift_alibi_for_padding(alibi, attention_mask, batch_size):
+    if jnp.isin(0, attention_mask):
+        unpadded_indices = nn.relu(lax.cumsum(attention_mask, axis=1) - 1)
```
If the purpose of the `nn.relu` is to perform a `max(0, x)` operation we can replace by `jax.lax.max`

patrickvonplaten committed:
update

patrickvonplaten committed:
remove bogus file

patrickvonplaten committed:
up

patrickvonplaten commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -153,21 +135,24 @@ def get_slopes_power_of_2(n):
+    query_length = 1
+
+    slopes = jnp.array(get_slopes(n_head))[None, :, None, None]
+    arange_tensor = attention_mask.cumsum(-1, dtype=dtype)[:, None, None, :] - 1
```
@younesbelkada - that's the trick we usually do for left padding :-)

patrickvonplaten committed:
get generation to work

patrickvonplaten committed:
clean code a bit

younesbelkada committed:
added small tests

younesbelkada committed:
adding albii test

younesbelkada committed:
make CI tests pass:

- change init weight
- add correct tuple for output attention
- add scan test
- make CI tests work

younesbelkada committed:
fix few nits

younesbelkada committed:
fix nit onnx

younesbelkada committed:
fix onnx nit

Commit:
add missing dtype args to nn.Modules

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18022). All of your documentation changes will be reflected on that endpoint.

Commit:
remove debugging statements

Commit:
fix scan generate

younesbelkada committed:
Merge branch 'main' into add_bloom_flax

younesbelkada committed:
Update modeling_flax_bloom.py

younesbelkada committed:
Update test_modeling_flax_bloom.py

younesbelkada committed:
Update test_modeling_flax_bloom.py

younesbelkada committed:
Update test_modeling_flax_bloom.py

younesbelkada committed:
fix small test issue + make style

patrickvonplaten commented:
Is this ready to merge @younesbelkada ? @sanchit-gandhi @patil-suraj could you take a final look?

younesbelkada commented:
Yeah ready to merge for me, for partitioning compatibility I think that we can adress another PR for that

younesbelkada commented:
Also we need to add the credits to @haileyschoelkopf for starting the contribution

Commit:
clean up

sanchit-gandhi commented on tests/models/bloom/test_modeling_flax_bloom.py:
```diff
@@ -0,0 +1,281 @@
+    from transformers.models.bloom.modeling_flax_bloom import build_alibi_tensor_flax
+
+
+def prepare_opt_inputs_dict(config, input_ids, attention_mask=None, head_mask=None):
```
Head mask not used
```suggestion
def prepare_bloom_inputs_dict(config, input_ids, attention_mask=None):
```

younesbelkada replied:
Thanks for the catch!

younesbelkada replied:
Also changed the function name in a test

sanchit-gandhi commented on tests/models/bloom/test_modeling_flax_bloom.py:
```diff
@@ -0,0 +1,281 @@
+        model = model_class_name(config)
+
+        input_ids = inputs_dict["input_ids"]
+        attention_mask = inputs_dict["attention_mask"]
```
This `attention_mask` gets overridden by the one on L132 (also in [`test_flax_opt`](https://github.com/huggingface/transformers/blob/2b81f72be9fa6d69734ae27cfcbfd72b04988fe4/tests/models/opt/test_modeling_flax_opt.py#L129), from which this was copied). Can remove this line.

younesbelkada replied:
Just removed the L132 and seems to work fine

sanchit-gandhi replied:
The one on L132 was superseded by a second definition of `attention_mask` 3/4 lines later, and so was un-used.

younesbelkada replied:
yeah so I removed the one on the L124 and directly the one from  `inputs_dict`

younesbelkada replied:
Sorry removed the wrong line actually, should be fixed in cecfcb158e7bfc2558d6f9f1ea7f270468e0dbc8

sanchit-gandhi reviewed (commented):
@younesbelkada some comments regarding the tests!

younesbelkada committed:
Update tests/models/bloom/test_modeling_flax_bloom.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

younesbelkada committed:
fix function name

sanchit-gandhi commented:
Modelling file and tests LGTM. Am in agreement that the the named axes should come as a follow-up PR - better to modify a script that's already in-place and working rather than merging a complete overhaul of how we structure our code in one go.

younesbelkada committed:
small fix test

patrickvonplaten commented:
Good to merge for me once @patil-suraj and @sanchit-gandhi give the green light here :-)

haileyschoelkopf committed:
forward contrib credits from PR17761

younesbelkada committed:
Fix failing test

younesbelkada committed:
fix small typo documentation

sanchit-gandhi reviewed (approved):
LGTM, thanks everyone! And special mention to @haileyschoelkopf who kicked-off this PR 🤗

For reference, the modelling code used in the deployment of Flax BLOOM is currently in a standalone repository: https://github.com/huggingface/bloom-jax-inference/blob/main/bloom_inference/modeling_bloom/modeling_bloom.py
It employs 2D model, activation and data sharding, and a faster generation method with more efficient indexing. These changes will be gradually brought across to the Transformers modelling code with the introduction of named axes support.

younesbelkada committed:
Merge branch 'main' into add_bloom_flax

younesbelkada committed:
fix non passing test

- remove device from build alibi

younesbelkada commented:
Thanks all for your comments! 
I had to fix a small test since the main modeling code has changed a bit in #18344 !
I think this PR is ready to be merged 💪 cc @patil-suraj

patrickvonplaten commented:
@patil-suraj @sanchit-gandhi - should we merge this one?

patrickvonplaten commented on src/transformers/modeling_flax_pytorch_utils.py:
```diff
@@ -109,8 +109,20 @@ def is_key_or_prefix_key_in_dict(key: Tuple[str]) -> bool:
 
 
 def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
+    try:
```
sounds good to me!

sanchit-gandhi replied:
Is this necessary? Looks to me to be redundant! We check for `torch` import in `load_pytorch_checkpoint_in_flax_state_dict`:
https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/modeling_flax_pytorch_utils.py#L45-L53
And then call `convert_pytorch_state_dict_to_flax` 9 lines later:
https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/modeling_flax_pytorch_utils.py#L62
So this check in `convert_pytorch_state_dict_to_flax`is superseded by the parent function `load_pytorch_checkpoint_in_flax_state_dict`.

younesbelkada replied:
Yeah let's remove it - I just had to add `# noqa: F821` to make the quality test pass here

patrickvonplaten commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+    use_scan: bool = False
+
+    # TODO(sanchit-gandhi): re-write as a `setup` to conform to Transformers JAX/Flax conventions
+    @nn.compact
```
@younesbelkada - this should be rewritten in with `def setup(...)` function as stated in the comment. Could you take care of it?

younesbelkada replied:
I should have proposed a fix in dda4a08e85680897dce646b5669603e7e7246c6e
But not sure if this breaks anything related to `use_scan`, will check later if there is a test for that

younesbelkada replied:
Managed to fix the test! 
I had to refactor a bit the code and use the following versions of `jax` and `flax`
```
jax.__version__
>>> ‘0.3.17’
flax.__version__
>>> ‘0.6.0’
```

patrickvonplaten reviewed (commented):
One small change regarding `setup` would be nice here as stated in the comment

sanchit-gandhi commented:
Good for me! Awaiting the green light from @patil-suraj who requested a look over the modelling code prior to merging. Regarding `setup`, I previously asked in the G Chat with the JAX/Flax guys how to make this work with `scan`. Think it went under the radar, so will bring this issue back up!

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+
+        attention_bias = attention_bias + alibi
+
+        # TODO(sanchit-gandhi): override softmax precision to fp32 if self.attention_softmax_in_fp32=True and self.dtype != fp32
```
Is this still standing ?
Here we could simply upcast the attention weights to `fp32` before doing softmax to achieve this.

younesbelkada replied:
I think yes, since it is still in place in the original modeling code. 
Proposed a simple fix in: c685442e3548dcf6c9ebaeb91a17a480eb683ba5

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+    def setup(self):
+        hidden_size = self.config.hidden_size
+
+        self.pretraining_tp = self.config.pretraining_tp
+        self.slow_but_exact = self.config.slow_but_exact
```
These two are not used
```suggestion
```

younesbelkada replied:
Fixed in 1999df780452ff4b0830c1d3eed49f098c083076 ! Could not managed to commit directly from your proposed change since the diff was outdated

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+
+        self.mlp = FlaxBloomMLP(self.config, dtype=self.dtype)
+
+        self.apply_residual_connection_post_layernorm = self.config.apply_residual_connection_post_layernorm
```
Is this option used ? When I was working on it all the model had the same value for `apply_residual_connection_post_layernorm`. If that's the case maybe we could remove this.

younesbelkada replied:
I think that we should keep it in case people train bloom models with different values for `apply_residual_connection_post_layernorm` so I would keep it I think for now

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+        deterministic: bool = True,
+        init_cache: bool = False,
+        output_attentions: bool = False,
+        layer_number: int = None,
```
`layer_number` is not used anywhere.

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+            deterministic=deterministic,
+            init_cache=init_cache,
+            output_attentions=output_attentions,
+            layer_number=layer_number,
```
`layer_number` is not used anywhere.

younesbelkada replied:
Should be fixed in fc371268add3de5cb68ad6ecb44cbbf6d9dde90a

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+    dtype: jnp.dtype = jnp.float32
+    use_scan: bool = False
+
+    # TODO(sanchit-gandhi): re-write as a `setup` to conform to Transformers JAX/Flax conventions
```
should be taken care of

younesbelkada replied:
Let's move the discussion at https://github.com/huggingface/transformers/pull/18022#discussion_r954115059

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+            self.config.vocab_size,
+            self.embed_dim,
+            embedding_init=embedding_init,
+            dtype=self.dtype,
```
We shouldn't pass the `dtype` to embeds, this will keep the embed weights in that `dtype`,  but it's fine if that's what we want here.

younesbelkada replied:
I think that this is what we want here 💪

sanchit-gandhi replied:
We should pass the `dtype` arg to keep the weights in bfloat16 for half precision inference.

Also related: https://github.com/huggingface/transformers/pull/18462

patil-suraj commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,771 @@
+        output_hidden_states: bool = False,
+        return_dict: bool = True,
+    ):
+        inputs_embeds = self.word_embeddings(input_ids.astype("i4"))
```
`input_ids` are already casted to `i4` in the pre-trained model class no ?

younesbelkada replied:
Some tests where failing when removing the casting, probably the `input_ids` are not casted to `i4` in the pre-trained model class - will investigate that

younesbelkada replied:
Thanks for pointing out this issue! Finally fixed in 05c05212fb629487227b5e20b16a57995b3709c7

patil-suraj reviewed (commented):
Thanks a lot for adding BLOOM in Flax !
LGTM, but some todo's are standing, should be good to merge once those are addressed.

younesbelkada committed:
refactor call

- refactor `FlaxBloomBlockCollection` module

younesbelkada committed:
Merge remote-tracking branch 'upstream2/main' into add_bloom_flax

younesbelkada committed:
make style

younesbelkada committed:
upcast to fp32

younesbelkada committed:
cleaner way to upcast

younesbelkada committed:
remove unused args

younesbelkada committed:
remove layer number

younesbelkada committed:
fix scan test

younesbelkada committed:
make style

younesbelkada committed:
fix i4 casting

younesbelkada committed:
Merge remote-tracking branch 'upstream2/main' into add_bloom_flax

younesbelkada commented:
Hi @sanchit-gandhi @patil-suraj @patrickvonplaten 
I think that the PR is ready for a final round of review! 💪  I proposed a potential solution to re-write the model definition with `setup` but I am unsure how this fits the HF+Jax approach.

patrickvonplaten commented:
@sanchit-gandhi @patil-suraj could you check here?

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,778 @@
+            for layer_number in range(self.config.num_hidden_layers)
+        ]
+
+        if self.use_scan:
```
Technically we don't actually need this `if` statement: `self.scan_fn` will only be instantiated if used in the `__call__` method.

younesbelkada replied:
Thanks! Should be fixed in b002f6494271dbaae5ee01ce0414c9e8d76435a1

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,778 @@
+                length=self.config.num_hidden_layers,
+            )(self.config, dtype=self.dtype, use_scan=True, name="FlaxBloomBlockLayers")
+
+    # TODO(sanchit-gandhi): re-write as a `setup` to conform to Transformers JAX/Flax conventions
```
We can remove this TODO
```suggestion
```

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,778 @@
+        hidden_states,
+        alibi,
+        attention_mask=None,
+        layer_past=None,
```
This arg defaults to `None` everywhere and is not used in any of the `__call__` methods - can we remove it @younesbelkada?

younesbelkada replied:
Yes we can! Fixed in 0b70bf22ee732fb741c2b2184164227d6b589470

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,778 @@
+                hidden_states,
+                alibi,
+                attention_mask,  # kwargs not supported by scan
+                None,
```
See here how `layer_past` is always set to `None` - if we trace it back to `FlaxBloomBlock` we see that it's not used anywhere in the BLOOM model.

sanchit-gandhi reviewed (commented):
Almost there! Just some small refactoring suggestions to clean the code up a bit

younesbelkada committed:
Update src/transformers/models/bloom/modeling_flax_bloom.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

younesbelkada committed:
fix slow test

younesbelkada committed:
Merge branch 'add_bloom_flax' of https://github.com/younesbelkada/transformers into add_bloom_flax

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add_bloom_flax

younesbelkada committed:
remove `layer_past`

younesbelkada committed:
refactor a bit

younesbelkada committed:
fix `scan` slow test

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,772 @@
+"""
+
+
+def build_alibi_tensor_flax(attention_mask, n_head, dtype):
```
This is currently quite difficult to understand with the triple nested function structure! Could we:
1) Add a docstring similar to [modeling_bloom.py#L89](https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/models/bloom/modeling_bloom.py#L89)
2) Unravel the triple nested function (copy the relevant logic from [modeling_bloom.py#L104-L118](https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/models/bloom/modeling_bloom.py#L104-L118))

younesbelkada replied:
Proposed a major refactor in 38c59465f44e52546b7b88c29aaf76429f87e94f !

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,772 @@
+
+        if self.use_scan:
+            # since all decoder layers are the same, we use nn.scan directly
+            # assert not output_attentions, "cannot use `scan` with `output_attentions` set to `True`"
```
We can uncomment these checks

younesbelkada replied:
Done in 38c59465f44e52546b7b88c29aaf76429f87e94f

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,772 @@
+        all_attentions = () if output_attentions else None
+        all_hidden_states = () if output_hidden_states else None
+
+        if self.use_scan:
```
We've currently left `scan` in the modelling code. Part of me thinks we should remove it for Transformers for the following reasons: 
1. `scan` adds a lot of boilerplate code that isn't very easy to understand
2. It is only beneficial for compile times when the model size is large, mostly when training and less so for inference

In the latter case, users will also likely shard the model, meaning they can employ the standalone code in [bloom-jax-inference](https://github.com/huggingface/bloom-jax-inference/blob/main/bloom_inference/modeling_bloom/modeling_bloom.py) where we can retain`scan`.

We also found generation time to be slower when using `scan` vs not using it (despite a faster compile time). The generation time will amortise the compile time in any use case of Flax BLOOM.

Given that the philosophy of Transformers is functional, easy-to-understand code that is not necessarily fully optimised, I'm in favour of stripping `scan` from Flax BLOOM and leaving it to [bloom-jax-inference](https://github.com/huggingface/bloom-jax-inference/blob/main/bloom_inference/modeling_bloom/modeling_bloom.py) to serve users that want to deploy larger variants of the model.

WDYT @patrickvonplaten @patil-suraj @younesbelkada?

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,772 @@
+    def setup(self):
+        self.embed_dim = self.config.hidden_size
+
+        embedding_init = jax.nn.initializers.normal(stddev=self.config.initializer_range)
```
Fine to use this value directly in the `nn.Embed` rather than separately defining a variable! No harm in what you've done though :)

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,772 @@
+            dropout_rate=self.config.attention_dropout,
+            deterministic=deterministic,
+            dtype=attention_dtype,
+            precision=None,
```
Can omit this arg as we've set it to its default value (`None`)

sanchit-gandhi commented on src/transformers/models/bloom/modeling_flax_bloom.py:
```diff
@@ -0,0 +1,772 @@
+        # do post-embedding layernorm
+        hidden_states = self.word_embeddings_layernorm(inputs_embeds)
+
+        batch_size, curr_seq_len, _ = hidden_states.shape
```
Don't see these variables being used

sanchit-gandhi reviewed (commented):
Thanks for addressing the previous comments @younesbelkada. Two things from this round of review:
1. Could we tidy up the `build_alibi_tensor_flax` function to avoid a triple nested function? Much of the logic can be copied over from PyTorch BLOOM!
2. Big question for me is whether we keep `scan` or not - I'm in favour of removing it for Flax BLOOM (see comments below)

younesbelkada committed:
remove useless import

younesbelkada committed:
major changes

- remove unused code
- refactor a bit
- revert import `torch`

younesbelkada committed:
major refactoring

- change build alibi

younesbelkada reviewed (commented):
Should have addressed your new suggestions @sanchit-gandhi ! Here I mainly focused on refactoring a bit the `build_alibi` function to match the implementational style of `Pytorch` -> this way it seems to be more readable !
I will leave you and @patil-suraj and @patrickvonplaten decide regarding the scan feature and happy to remove it once we agree on that!
Can also confirm the slow tests/conversion tests pass ;)

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add_bloom_flax

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add_bloom_flax

younesbelkada commented:
Should we merge this one ? cc @patrickvonplaten @patil-suraj @sanchit-gandhi

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18022). All of your documentation changes will be reflected on that endpoint.

wangkuiyi commented:
Both https://github.com/huggingface/transformers/pull/17761 and this PR show a lot of work. Why isn't it merged?

sanchit-gandhi commented:
Adding this to my TODOs
